src/onboarding.ts

```diff
@@ -228,14 +229,66 @@ promptfoo eval
 Afterwards, you can view the results by running \`promptfoo view\`
 `;
 
-function recordOnboardingStep(step: string, properties: Record<string, EventValue> = {}) {
+function recordOnboardingStep(step: string, properties: any = {}) {
   telemetry.recordAndSend('funnel', {
     type: 'eval onboarding',
     step,
     ...properties,
   });
 }
 
+/**
+ * Iterate through user choices and determine if the user has selected a provider that needs an API key
+ * but has not set and API key in their enviorment.
+ */
+export function reportProviderAPIKeyWarnings(providerChoices: (string | object)[]): string[] {
+  // Dictionary of what warnings may have already been printed for the user.
+  const shownWarnings = new Map([
```

Comment:
can you generate this map dynamically instead of hardcoding it? So if we add a new provider to `ProvidersRequiringAPIKeysEnvKeys` we don't have to update this code?

Reply:
Hmmm I think this could work but only a very small subset of providers (I am told) require API keys. Given that I think we would end up with a bunch of code and large and not-readable object in code to only support two providers which require keys. I think a bigger and better refactor here would be to establish a larger class in envVars that ties each provider in the envVars type to an extensible interface that we can attach arbitrary information to such that, during onboarding, we can give the user info about how their onboarding selection is affected by their environment. I think that particular refactor is outside the scope of this change though. I can change it if you feel strongly about it but I think this is more readable and easier to understand what (small) subset of providers would actually trigger this warning while also reducing mental overhead when someone comes to inevitably read + modify this function. lmk what you think

Reply:
Updated!